cocos/core/renderer/core/program-lib.ts

```diff
@@ -442,9 +443,28 @@ class ProgramLib {
         const tmpl = this._templates[name];
         const tmplInfo = this._templateInfos[tmpl.hash];
         if (!tmplInfo.pipelineLayout) {
-            this.getDescriptorSetLayout(device, name); // ensure set layouts have been created
-            insertBuiltinBindings(tmpl, tmplInfo, globalDescriptorSetLayout, 'globals');
-            tmplInfo.setLayouts[SetIndex.GLOBAL] = pipeline.descriptorSetLayout;
+            const director = legacyCC.director;
+            if (director.root.usesCustomPipeline) {
+                const root = legacyCC.director.root;
+                const ppl: Pipeline = root.customPipeline;
+                insertBuiltinBindings(tmpl, tmplInfo, globalDescriptorSetLayout, 'globals');
+                const globalDS = ppl.getDescriptorSetLayout(name, UpdateFrequency.PER_PASS);
+                const materialDS = ppl.getDescriptorSetLayout(name, UpdateFrequency.PER_BATCH);
+                const localDS = ppl.getDescriptorSetLayout(name, UpdateFrequency.PER_INSTANCE);
+                if (globalDS) {
+                    tmplInfo.setLayouts[SetIndex.GLOBAL] = globalDS;
+                }
+                if (materialDS) {
+                    tmplInfo.setLayouts[SetIndex.MATERIAL] = materialDS;
+                }
+                if (localDS) {
+                    tmplInfo.setLayouts[SetIndex.LOCAL] = localDS;
+                }
+            } else {
```

Comment:
Adding custom pipeline, refactoring ProgramLib

Reply:
Should also modify native implementation?

Reply:
Yes, native implementation will be added later. Currently custom pipeline cannot be enabled yet, so this code path will not be reached yet.

Reply:
Got it.